src/native/eventpipe/ep-session.c

```diff
@@ -34,6 +34,19 @@ ep_session_remove_dangling_session_states (EventPipeSession *session);
  * EventPipeSession.
  */
 
+static size_t streaming_loop_tick(EventPipeSession *const session, bool *events_written) {
+	if (!ep_session_get_streaming_enabled (session)){
+		session->streaming_thread = NULL;
+		return 1; // done
+	}
+	if (!ep_session_write_all_buffers_to_file (session, events_written)) {
+		session->streaming_thread = NULL;
+		ep_disable ((EventPipeSessionID)session);
+		return 1; // done
+	}
+	return 0; // try again
+}
+
 EP_RT_DEFINE_THREAD_FUNC (streaming_thread)
```

Comment:
Maybe it is simpler and more clear to just create two functions keeping the first error/state checking code and then call into another function implemented differently for single and multi threaded support, you can still call the loop tick, but it might just be as simple to inline that part into each implementation once you have two methods.